src/stateMachine/mutation/BQL/enrichSteps/preHookDependencies.ts

```diff
@@ -0,0 +1,287 @@
+/* eslint-disable no-param-reassign */
+import { produce } from 'immer';
+import { traverse } from 'object-traversal';
+import { isObject } from 'radash';
+import type {
+	BQLField,
+	BormConfig,
+	DBHandles,
+	EnrichedBQLMutationBlock,
+	EnrichedBormSchema,
+	FilledBQLMutationBlock,
+} from '../../../../types';
+import { getCurrentSchema } from '../../../../helpers';
+import { queryPipeline } from '../../../../pipeline/pipeline';
+import { DBNode } from '../../../../types/symbols';
+
+export const preQueryPathSeparator = '___';
+type ObjectPath = { beforePath: string; ids: string | string[]; key: string };
+
+export const preHookDependencies = async (
+	blocks: EnrichedBQLMutationBlock | EnrichedBQLMutationBlock[],
+	schema: EnrichedBormSchema,
+	config: BormConfig,
+	dbHandles: DBHandles,
+) => {
+	const getFieldKeys = (block: FilledBQLMutationBlock | Partial<FilledBQLMutationBlock>, noDataFields?: boolean) => {
+		return Object.keys(block).filter((key) => {
+			if (!key.startsWith('$') && block[key] !== undefined) {
+				if (noDataFields) {
+					const currentSchema = getCurrentSchema(schema, block);
+					if (currentSchema.dataFields?.find((field) => field.path === key)) {
+						return false;
+					} else {
+						return true;
+					}
+				} else {
+					return true;
+				}
+			} else {
+				return false;
+			}
+		});
+	};
+
+	if (!blocks) {
+		throw new Error('[BQLE-M-PQ-1] No blocks found');
+	}
+	/// 1. Convert the mutation to a query to get all $fields
+	const convertMutationToQuery = (blocks: FilledBQLMutationBlock[]) => {
+		const processBlock = (block: FilledBQLMutationBlock, root?: boolean) => {
+			let $fields: any[] = block.$fields || [];
+			const filteredBlock = {};
+			const toRemoveFromRoot = ['$op', '$bzId', '$parentKey'];
+			const toRemove = ['$relation', '$entity', '$id', ...toRemoveFromRoot];
+			for (const k in block) {
+				if (toRemoveFromRoot.includes(k)) {
+					continue;
+				}
+				if (toRemove.includes(k) && !root) {
+					continue;
+				}
+				if (!k.includes('$') && (isObject(block[k]) || Array.isArray(block[k]))) {
+					const v = block[k];
+					if (Array.isArray(v) && v.length > 0) {
+						v.forEach((opBlock) => {
+							const newField = {
+								$path: k,
+								...processBlock(opBlock),
+							};
+							if ($fields.find((o) => o.$path === newField.$path)) {
+								// do nothing
+							} else if ($fields.find((o) => o === newField.$path)) {
+								const filteredFields = $fields.filter((o) => o !== newField.$path);
+								$fields = [...filteredFields, ...[newField]];
+							} else {
+								$fields = [...$fields, ...[newField]];
+							}
+						});
+					} else {
+						const newField = {
+							$path: k,
+							...processBlock(v),
+						};
+						$fields = [...$fields, ...[newField]];
+					}
+				} else {
+					// @ts-expect-error todo
+					filteredBlock[k] = block[k];
+				}
+			}
+			return {
+				...filteredBlock,
+				$fields,
+			};
```

Comment:
We don't need to create a new object. We can set `$fields` to filteredBlocks` and return it.